lib/ruleHelper.js

```diff
@@ -76,15 +77,77 @@ RuleHelper.prototype = {
-            allowed = this.allowedExpression(expression.expression, escapeObject);
+            allowed = this.allowedExpression(expression.expression, escapeObject, details);
         } else if (expression.type === "TypeCastExpression") {
-            allowed = this.allowedExpression(expression.expression, escapeObject);
+            allowed = this.allowedExpression(expression.expression, escapeObject, details);
         } else if (Array.isArray(expression)) {
-            allowed = expression.every((e) => this.allowedExpression(e, escapeObject));
+            allowed = expression.every((e) => this.allowedExpression(e, escapeObject, details));
+        } else if (expression.type === "Identifier") {
+            // find declared variables and see which are literals
+            const scope = this.context.getScope(expression);
+            const variableInfo = scope.set.get(expression.name);
+
+            // let's see if this variable is allowed, by tracing it back through the code
+            if (variableInfo) {
+                // look if the var was defined as allowable
+                let definedAsAllowed = false;
+                for (const def of variableInfo.defs) {
+                    if (def.node.type === "VariableDeclarator") {
+                        if ((def.kind !== "let") && (def.kind !== "const")) {
+                            // We do not allow for identifiers declared with "var", as they can be overridden in a
+                            // way that is hard for us to follow (e.g., assignments to globalThis[theirNameAsString]).
+                            continue;
+                        }
+
+                        // the `init` property carries the right-hand side of the variable definition:
+                        const varInitAs = def.node.init;
+                        if (!this.allowedExpression(varInitAs, escapeObject, details)) {
+                            // if one variable definition is considered unsafe, all are.
+                            // NB: order of definition is unclear. See issue #168.
+                            if (!details.message) {
+                                const {line, column} = varInitAs.loc.start;
+                                details.message = `Variable '${expression.name}' initialized with unsafe value at ${line}:${column}`;
+                            }
+                            definedAsAllowed = false;
+                            break;
+                        }
+
+                        // keep iterating through other definitions.
+                        definedAsAllowed = true;
+                    }
+                }
+                if (definedAsAllowed) {
+                    // the variable was declared as a safe value (e.g., literal)
+                    // now inspect writing references to that variable
+                    let allWritingRefsAllowed = false;
+                    for (const ref of variableInfo.references) {
```

Comment:
Not sure if it is possible in practice, but if `variableInfo.references` is an empty array, then iterating over `variableInfo.defs` would be pointless because we will still consider the variable as unsafe, and so it may be reasonable to check up front `variableInfo?.defs.length > 0 && variableInfo?.references.length > 0` and return earlier if it is false (without even trying to track back the identifier).

Reply:
Heh, that optional chaining isn't supported, so now I had to bump eslint and change the config. This PR is growing in size :)

Reply:
Edit: We (currently) support very old node. I don't want to address the question of nodejs in this exact pull request, but optional chaining requires 13 and above. I'll revert and write code that doesn't require it.